File: sortingAlgo/mergeSort/mergeSort.c
#include<stdio.h>

void merge(int A[], int i1, int j1, int i2, int j2){
	int temp[j2 -i1 + 5];
	int i = i1, j = i2, k = 0;
	
	while(i<=j1 && j<=j2)	
	{
		if(A[i]<A[j])
			temp[k++]=A[i++];
		else
			temp[k++]=A[j++];
	}

	while(j<=j2)	
		temp[k++]=A[j++];

	while(i<=j1)	
		temp[k++]=A[i++];

	for(i=i1,j=0;i<=j2;i++,j++)
		A[i]=temp[j];
}


void mergesort( int Arr[], int i, int j){

	int mid;

	if(i<j){
		mid=(i+j)/2;
		mergesort(Arr, i , mid);
		mergesort(Arr, mid + 1, j);
		merge(Arr, i, mid, mid+1, j);
	}
}

int main(){

	int Arr[] = {345,5435,67,356,6547,74};

	mergesort(Arr, 0, 5);

	for(int i=0 ; i<6 ; i++ ){
		printf("%d ",Arr[i]);
	}
	return 0;
}

File: sortingAlgo/mergeSort/MergeSort.java
class Hacktoberfest
{

   static void mergeS(int arr[],int l,int m,int r)
   {
      int n1 = m - l + 1;
      int n2 = r - m;

      int A[] = new int [n1]; // Create temp arrays 
      int B[] = new int [n2]; // Create temp arrays 

      
      for (int i=0; i<n1; ++i){
         A[i] = arr[l + i];
      }
      for (int j=0; j<n2; ++j){
         B[j] = arr[m + 1+ j];
      }
      int i = 0, j = 0;
      int k = l;
      while (i < n1 && j < n2)
      {
         if (A[i] <= B[j])
         {
            arr[k] = A[i];
            i++;
         }
         else
         {
            arr[k] = B[j];
            j++;
         }
         k++;
      }

      /* Copy remaining elements of A[] if any */
      while (i < n1)
      {
         arr[k] = A[i];
         i++;
         k++;
      }

      /* Copy remaining elements of B[] if any */
      while (j < n2)
      {
         arr[k] = B[j];
         j++;
         k++;
      }
   }

   static void merge(int array[], int left, int right)
   {
      if (left < right)
      {
         
         int mid = (left+right)/2;//the middle point

         merge(array,left,mid); // Divide into two half
         merge(array,mid+1,right);

         mergeS(array,left,mid,right); // merging Two Sorted Half
      }
   }

   public static void main(String args[])
   {
      int array[] = {2,5,6,4,3,2,1};
      int left=0;
      int n=array.length-1;
      merge(array,left,n);

      // Printing Array
      for (int i = 0; i <= n; ++i)
      {
         System.out.print(array[i]+" ");
      }
      System.out.println();
   }
}

